src/Tasks/SystemState.cs

```diff
@@ -544,12 +549,73 @@ out fileState.frameworkName
         }
 
         /// <summary>
-        /// Cached implementation of GetDirectories.
+        /// Reads in cached data from stateFiles to build an initial cache. Avoids logging warnings or errors.
         /// </summary>
-        /// <param name="path"></param>
-        /// <param name="pattern"></param>
-        /// <returns></returns>
-        private string[] GetDirectories(string path, string pattern)
+        internal static SystemState DeserializePrecomputedCaches(ITaskItem[] stateFiles, TaskLoggingHelper log, Type requiredReturnType, GetLastWriteTime getLastWriteTime, AssemblyTableInfo[] installedAssemblyTableInfo, Func<string, bool> fileExists)
+        {
+            SystemState retVal = new SystemState();
+            retVal.SetGetLastWriteTime(getLastWriteTime);
+            retVal.SetInstalledAssemblyInformation(installedAssemblyTableInfo);
+            retVal.isDirty = stateFiles.Length > 0;
+            HashSet<string> assembliesFound = new HashSet<string>();
+            fileExists ??= FileSystems.Default.FileExists;
+
+            foreach (ITaskItem stateFile in stateFiles)
+            {
+                // Verify that it's a real stateFile; log message but do not error if not
+                SystemState sysState = DeserializeCache<SystemState>(stateFile.ToString(), log, false);
+                if (sysState == null)
+                {
+                    continue;
+                }
+                foreach (KeyValuePair<string, FileState> kvp in sysState.instanceLocalFileStateCache)
+                {
+                    string relativePath = kvp.Key;
+                    if (!assembliesFound.Contains(relativePath))
+                    {
+                        FileState fileState = kvp.Value;
+                        string fullPath = Path.GetFullPath(Path.Combine(Path.GetDirectoryName(stateFile.ToString()), relativePath));
+                        if (fileExists(fullPath))
+                        {
+                            // Correct file path and timestamp
+                            fileState.LastModified = retVal.getLastWriteTime(fullPath);
```

Comment:
Is this the right timestamp? Doesn't it mean that we would potentially get incorrect information if the user updated one of the cached assemblies before a clean build?

Reply:
Yes! Thank you for pointing that out. I removed that line entirely, and it would have made the cache (often) wrong. This was good behavior when I was checking for mvids but is no longer correct, and I didn't remove it.